### ISSUE TITLE
api/compare: return basic information about the *actual* baseline (vs requested)

### ISSUE THREAD

**Issue by austin3dickey** [conbench/conbench#1030]
Currently, for the lookback z-score analysis that happens when hitting the `/api/compare/*/baseline_id...contender_id` endpoints, we compare the contender results to a distribution of results ending with the baseline commit. But what if there were no matching results associated with the baseline commit? It would be helpful to `benchalerts` (see #969) to report the "actual" baseline commit used for this analysis; that is, the most-recent commit in history that was used. If possible, we should return the number of "skipped" commits too.

**Comment by jonkeane** [comment]
One thing that's always been a little bit funny about these endpoints is that they are serve two slightly different purposes depending on which one we are talking about: 

* `compare/runs/...` This gives you a list of benchmarkresults that have matches in both runs. There's no real guessing or redirection here. Either the benchmarkresults exist in both runs or they don't  
* `compare/batches/...` Does very similar to runs, just at the batch level (e.g. a small handful of benchmarkresults)
* `compare/benchmarks/...` Compares two benchmark results directly. Typically these match, but you could in principle actually compare across different ones (so long as you construct the URL yourself right now https://conbench.ursa.dev/compare/benchmarks/53cfcb86f6034973aa8baf15662fef73...d2fc1b7cbc794a52afa1547b0434a9cb/). But you get the comparison that you ask for.
* `compare/commits/...` Is the odd one. This doesn't actually give you a comparison, but rather it gives a set of _possible_ comparisons that one could then iterate over e.g. https://conbench.ursa.dev/api/compare/commits/0e1f5c6c2ce7504d86dd10d91ab7d29c2441d4ce...efb0be0d146715663fbfe9162f91da2e8e05b58f/ notably this is also the only one of these that doesn't have a corresponding UI page (it's API only).

Ultimately, there are two activities here: doing a comparison and finding reasonable comparisons to do. All but `compare/commits/...` do the comparison given, where as `compare/commits/...` is finding reasonable comparisons to do and giving you some URLs where those comparisons could be found. We should probably split these two use cases up (or at least document them much better than they are now, since it's confusing!)

The thrust of this issue though is IMO really only something we should do for the finding reasonable comparisons to do use case only. The _doing the comparison_ use case should give you what you ask for even if it's absurd or results in `null` (and only once we have that entirely sorted + are doing it correctly, maaaaaybe we consider sending someone to a disambiguation page or something — but this is not a requirement or feature we need to support right now).

**Comment by austin3dickey** [comment]
Ah, great callout. Like you mentioned in https://github.com/conbench/conbench/issues/384#issuecomment-1494822526, I've been thinking about how we could remove the "finding reasonable comparisons" use case, at least from the `/api/compare` namespace. I was going to propose this:

- `/api/compare/benchmark-results/baseline_result...contender_result`
  - pairwise compare `baseline_result` and `contender_result`
  - do a lookback z-score analysis on `contender_result` with a distribution ending at `baseline_result`'s commit
- `/api/compare/batches/baseline_batch...contender_batch`
  - find all results belonging to `baseline_batch`
  - find all results belonging to `contender_batch`
  - do a FULL OUTER JOIN ON pairs that share the same case/context/hardware/repo
  - for each pair, compare them as in `/api/compare/benchmark-results/`
- Then `/api/compare/runs/` and `/api/compare/commits/` behave like `/api/compare/batches/`.

For something like `/api/compare/runs/` this could return more than one comparison item for each contender result, but we'd also signify the one to alert on (#955).

**Comment by austin3dickey** [comment]
We chatted about this in person, and decided it was worth it to keep around both use cases of "get me a comparison between these two IDs" and "give me a list of valid URIs to get those comparisons." I think a practical proposal is the following:

- `/api/compare/benchmark-results/baseline_id...contender_id/`
  - Returns a dict with info, stats, and analyses (schema to be determined later)
  - Always returns this comparison, even if the results are not "comparable" (i.e. they don't share case/context/hardware/repo)
- `/api/compare/runs/baseline_id...contender_id/`
  - Returns a list of "comparable" comparisons (as dicts with the same schema as above)
  - Each result on the contender run is represented in 0 or 1 of these comparison-dicts: never more than 1, and only 0 if there are no "comparable" results on the baseline run
  - Might return an empty list, which signifies there are no results on the contender run or no "comparable" results on the baseline run
- `/api/compare/commits/contender_hash/?target_baseline_commit={parent,fork_from_default,head_of_default}[&repository=ABC]`
  - Returns metadata + a list of URIs to the `/api/compare/runs/` endpoint that are each guaranteed to generate a non-empty list of comparisons
  - Each run on the contender commit is represented in 0 or 1 of these URIs: never more than 1
  - When choosing the baseline run for each run on the contender commit, start at the runs on the target baseline commit, and work backwards in git history until we find one that has comparisons; if there are none, a URI will not be returned for that run
  - I prefer to search by `contender_hash` (and optional `repository`) over `commit_id`, because that's already known on the client side

Note this removes the `/batches` endpoint. That UI page would probably go away.

**Comment by jonkeane** [comment]
Yeah, this sounds good.

> /api/compare/runs/baseline_id...contender_id/
> * Returns a list of "comparable" comparisons (as dicts with the same schema as above)
> * Each result on the contender run is represented in 0 or 1 of these comparison-dicts: never more than 1, and only 0 if there are no "comparable" results on the baseline run
> * Might return an empty list, which signifies there are no results on the contender run or no "comparable" results on the baseline run

I'm not 100% sure what the behavior is today WRT "we only return results if they match" but I wonder if we should actually return everything here (things that are in the baseline only, things that are in the contender only, and things that match across) and then filter in the client | UI to be only the ones that have comparisons. This could of course be guarded by a flag or something too if we wanted. This way consumers of this API could more easily represent things like "this run has a bunch of things that aren't matched with the baseline" or possibly let folks take two disjoint things (e.g. they vary in case value) and compare them (granted at this point it would be offline at the client level, but that is ok for now) if they knew that that was a comparison that they want + makes sense for their usecase.


> /api/compare/commits/contender_hash/?target_baseline_commit={parent,fork_from_default,head_of_default}[&repository=ABC]
This is a little pedantic, but I would keep this bullet point separate (maybe even a totally separate issue). It is conceptually different from the activity of comparing and keeping it close by + in lists like this IMO is confusing.

> I prefer to search by contender_hash (and optional repository) over commit_id, because that's already known on the client side

Could you say more here? are you thinking that ` /api/compare/commits/contender_hash/?target_baseline_commit={parent,fork_from_default,head_of_default}[&repository=ABC]` should _also_ accept commit hash as well as the aliases "parent", "fork_from_default", "head_of_default"? If so: yup, I 💯 agree!

**Comment by austin3dickey** [comment]
>This is a little pedantic, but I would keep this bullet point separate (maybe even a totally separate issue)

Good point. I originally included everything here because this ticket was created to discuss how to handle the case when you find a comparison on a different commit from the one you targeted. But we've decided that's not really what we'll do: instead we'll have "a way to get URIs to good comparisons" and "a way to get those comparison results" -- which are two separate designs.

I think I'll close this ticket as "not planned" and open a new set of tickets for the new plan (linking to this one for record-keeping).
